### PR TITLE
Added Use of Cone Vision to OneFivePlusBump and OneFiveBumpBalance

### PR DESCRIPTION
**Pull request by avakimc**

### CODE REVIEW

**Review by tails618 — CHANGES_REQUESTED** [review]
Rename "...BumpFirstHalf" to "...Bump" and rename "...BumpSecondHalf" to "...BumpReturn" (Make sure to keep "Red" for the red ones)

File: src/main/java/frc/robot/commands/autos/OneFiveBalanceBump.java [hunk 1]
```diff
@@ -3,86 +3,103 @@
...
+            }),
+        new ParallelDeadlineGroup(
+            new SequentialCommandGroup(
+                new WaitCommand(0.5),
+                new RunCommand(
+                        () -> {
+                          drive.rotateOrKeepHeading(NORM_SPEED_INTAKING, 0, 0, false, -1);
+                        })
+                    .until(
+                        () -> (drive.getPose().getX() - startXMeters) > DISTANCE_AT_CONE_METERS)),
```
Comment (balghane) [inline src/main/java/frc/robot/commands/autos/OneFiveBalanceBump.java]: Can we use the new DriveDistance command?

File: src/main/java/frc/robot/commands/autos/OneFiveBalanceBump.java [hunk 2]
```diff
@@ -3,86 +3,103 @@
...
+                .finallyDo(
+                    (boolean interrupted) -> {
+                      lift.home();
+                      lift.closeGrabber();
+                      intake.setDesiredDeployed(false);
+                      intake.setDesiredClamped(false);
+                      intake.stopIntakingGamePiece();
+                    })),
+        new SwerveToPointWrapper(red, drive, new Pose2d (5.85, 2.5, Rotation2d.fromDegrees(0.0)), 2.0),
         new ChargeFarSide(drive, DISTANCE_ONTO_CHARGE_STATION_METERS));
```
Comment (balghane) [inline src/main/java/frc/robot/commands/autos/OneFiveBalanceBump.java]: Can we use the new `DriveUntilBalanced` in the backwards direction?

File: src/main/java/frc/robot/commands/autos/OneFivePlusBump.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import frc.robot.subsystems.Lights;
+import frc.robot.subsystems.drive.DriveSubsystem;
+import frc.robot.utils.ScoringLocationUtil;
+
+/**
+ * Assuming the robot is starting from cone scoring position nearest the loading zone, this auto
+ * component drives to the nearest midfield game object, picks it up, and scores it on the next
+ * nearest cone position.
+ */
+public class OneFivePlusBump extends SequentialCommandGroup {
```
Comment (balghane) [inline src/main/java/frc/robot/commands/autos/OneFivePlusBump.java]: Fix class comment, maybe rename OneFiveBumpReturn?

File: src/main/java/frc/robot/commands/autos/OneFivePlusBump.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            }),
+        new ParallelDeadlineGroup(
+            new SequentialCommandGroup(
+                new WaitCommand(0.5),
+                new RunCommand(
+                        () -> {
+                          drive.rotateOrKeepHeading(NORM_SPEED_INTAKING, 0, 0, false, -1);
+                        })
+                    .until(
+                        () -> (drive.getPose().getX() - startXMeters) > DISTANCE_AT_CONE_METERS)),
```
Comment (balghane) [inline src/main/java/frc/robot/commands/autos/OneFivePlusBump.java]: I think we need to drive back after the `ParallelDeadlineGroup`, then run the second half trajectory

Comment (balghane) [reply src/main/java/frc/robot/commands/autos/OneFivePlusBump.java]: Also same comment as the other command about using DriveDistance

File: src/main/java/frc/robot/RobotContainer.java [hunk 5]
```diff
@@ -101,10 +101,10 @@ public void initialize() {
         new JustTwoGamePieces(true, lift, intake, drive, lights, tagLimelight, scoreLoc));
     autonChooser.addOption(
         "1.5 balance bump blue",
-        new OneFiveBalanceBump(false, false, lift, intake, drive, lights, tagLimelight, scoreLoc));
+        new OneFiveBalanceBump(false, false, lift, intake, drive, lights, intakeLimelight, scoreLoc));
     autonChooser.addOption(
         "1.5 balance bump red",
-        new OneFiveBalanceBump(true, false, lift, intake, drive, lights, tagLimelight, scoreLoc));
+        new OneFiveBalanceBump(true, false, lift, intake, drive, lights, intakeLimelight, scoreLoc));
     autonChooser.addOption(
```
Comment (balghane) [inline src/main/java/frc/robot/RobotContainer.java]: Can you add the two 1.5 bump return autos?

File: src/main/java/frc/robot/commands/autos/OneFiveBalanceBump.java [hunk 6]
```diff
@@ -3,86 +3,100 @@
...
+        new InstantCommand(
+            () -> {
+              startXMeters = drive.getPose().getX();
+            }),
+        drive.followTrajectoryCommand(firstTraj, true),
+        // Turn to cone, intake it
+        new InstantCommand(
+            () -> {
+              drive.offsetCurrentHeading(limelight.getAngleToConeDeg());
+            }),
```
Comment (balghane) [inline src/main/java/frc/robot/commands/autos/OneFiveBalanceBump.java:78]: I think we need to add one of these to let the robot rotate
```
        new RunCommand(
                () -> {
                  drive.rotateOrKeepHeading(0, 0, 0, true, -1);
                })
            .withTimeout(0.25),
```

File: src/main/java/frc/robot/commands/autos/OneFiveBalanceBump.java [hunk 7]
```diff
@@ -3,86 +3,100 @@
...
+        new ParallelDeadlineGroup(
+            new SequentialCommandGroup(
+                new WaitCommand(0.5),
+                new DriveDistance(drive, NORM_SPEED_INTAKING, 1.5, 0.0, false)),
```
Comment (balghane) [inline src/main/java/frc/robot/commands/autos/OneFiveBalanceBump.java]: Put 1.5 meters into a named variable

File: src/main/java/frc/robot/commands/autos/OneFiveBalanceBump.java [hunk 8]
```diff
@@ -3,86 +3,100 @@
...
+            new IntakeSequence(intake, lift, lights)
+                .finallyDo(
+                    (boolean interrupted) -> {
+                      lift.home();
+                      lift.closeGrabber();
+                      intake.setDesiredDeployed(false);
+                      intake.setDesiredClamped(false);
+                      intake.stopIntakingGamePiece();
+                    })),
+        new SwerveToPointWrapper(red, drive, new Pose2d (5.85, 2.5, Rotation2d.fromDegrees(0.0)), 2.0),
```
Comment (balghane) [inline src/main/java/frc/robot/commands/autos/OneFiveBalanceBump.java]: Can you make this Pose2d a variable?

File: src/main/java/frc/robot/commands/autos/OneFiveBumpReturn.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * Assuming the robot is starting from cone scoring position nearest the loading zone, this auto
+ * component drives to the nearest midfield game object, picks it up, and returns to starting
+ * position
+ */
+public class OneFiveBumpReturn extends SequentialCommandGroup {
+  private static final double DISTANCE_AT_CONE_METERS = 5.9;
+  private static final double NORM_SPEED_INTAKING = 0.3;
+  private double startXMeters = 0;
```
Comment (balghane) [inline src/main/java/frc/robot/commands/autos/OneFiveBumpReturn.java]: I don't think we need this variable anymore for either of these autos, same for `DISTANCE_AT_CONE_METERS `

File: src/main/java/frc/robot/commands/autos/OneFiveBumpReturn.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        new InstantCommand(
+            () -> {
+              startXMeters = drive.getPose().getX();
+            }),
+        drive.followTrajectoryCommand(firstTraj, true),
+        // Turn to cone, intake it
+        new InstantCommand(
+            () -> {
+              drive.offsetCurrentHeading(limelight.getAngleToConeDeg());
+            }),
```
Comment (balghane) [inline src/main/java/frc/robot/commands/autos/OneFiveBumpReturn.java:88]: Also need to allow the robot to rotate here

File: src/main/java/frc/robot/commands/autos/OneFiveBumpReturn.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        new ParallelDeadlineGroup(
+            new SequentialCommandGroup(
+                new WaitCommand(0.5),
+                new DriveDistance(drive, NORM_SPEED_INTAKING, 1.5, 0.0, red),
+                new IntakeSequence(intake, lift, lights)
```
Comment (balghane) [inline src/main/java/frc/robot/commands/autos/OneFiveBumpReturn.java]: I don't think this parallel deadline group is done right

**Pull request closed without merging**